poetry/mixology/incompatibility.py

```diff
@@ -57,7 +57,8 @@ def __init__(
                     # irrelevant, since we already know that mutually exclusive version
                     # ranges are incompatible. We should never derive an irrelevant
                     # incompatibility.
-                    assert by_ref[ref] is not None
+                    if by_ref[ref] is None:
+                        raise AssertionError
```

Comment:
Hello @pnijhara,

thanks a lot for your contribution and trying to make the code better. However, I don't agree with the suggested changes here.

Using `assert` outside a test is in my opinion absolutely valid. They can be used as an internal self-check. Usually they should never raise, but if so you know there is something wrong in your code. The change you suggest here is just another way of writing the same thing.

Reply:
i don't now whether this is the intention, but the changes would also raise errors when the interpreter would be called with the [-O option](https://docs.python.org/3/using/cmdline.html#cmdoption-o).

Reply:
Well, that's the intention. As far as I know, while running python with -O (optimize) flag, it will remove the assert statements. So, it is a good practice to not to use assert statements outside of tests IMHO.

Reply:
@pnijhara AFAIK asserts are also slow. Maybe you could run a small performance test as an additional argument?

Reply:
@mvoitko I am not sure about the speed. But yes, I can test it with `timeit`.